src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdServerConfiguration.cs

```diff
@@ -0,0 +1,247 @@
+            {
+                options.UserinfoEndpointPath = "/OrchardCore.OpenId/UserInfo/Me";
+            }
+            if (settings.AllowAuthorizationCodeFlow)
+            {
+                options.GrantTypes.Add(OpenIdConnectConstants.GrantTypes.AuthorizationCode);
+            }
+            if (settings.AllowClientCredentialsFlow)
+            {
+                options.GrantTypes.Add(OpenIdConnectConstants.GrantTypes.ClientCredentials);
+            }
+            if (settings.AllowImplicitFlow)
+            {
+                options.GrantTypes.Add(OpenIdConnectConstants.GrantTypes.Implicit);
+            }
+            if (settings.AllowPasswordFlow)
+            {
+                options.GrantTypes.Add(OpenIdConnectConstants.GrantTypes.Password);
+            }
+            if (settings.AllowRefreshTokenFlow)
+            {
+                options.GrantTypes.Add(OpenIdConnectConstants.GrantTypes.RefreshToken);
+            }
+
+            options.Scopes.Add(OpenIdConnectConstants.Scopes.Email);
+            options.Scopes.Add(OpenIdConnectConstants.Scopes.Phone);
+            options.Scopes.Add(OpenIdConnectConstants.Scopes.Profile);
+            options.Scopes.Add(OpenIddictConstants.Claims.Roles);
+        }
+
+        public void Configure(OpenIddictOptions options) => Debug.Fail("This infrastructure method shouldn't be called.");
+
+        public void Configure(string name, JwtBearerOptions options)
+        {
+            // Ignore JWT handler instances that don't correspond to the private instance managed by the OpenID module.
+            if (!string.Equals(name, OpenIdConstants.Schemes.Userinfo, StringComparison.Ordinal))
+            {
+                return;
+            }
+
+            var settings = GetServerSettingsAsync().GetAwaiter().GetResult();
+            if (settings == null)
+            {
+                return;
+            }
+
+            options.TokenValidationParameters.ValidAudience = OpenIdConstants.Prefixes.Tenant + _shellSettings.Name;
+            options.TokenValidationParameters.IssuerSigningKeys = _serverService.GetSigningKeysAsync().GetAwaiter().GetResult();
+
+            // If an authority was explicitly set in the OpenID server options,
+            // prefer it to the dynamic tenant comparison as it's more efficient.
+            if (!string.IsNullOrEmpty(settings.Authority))
+            {
+                options.TokenValidationParameters.ValidIssuer = settings.Authority;
+            }
+            else
+            {
+                options.TokenValidationParameters.IssuerValidator = (issuer, token, parameters) =>
+                {
+                    if (!Uri.TryCreate(issuer, UriKind.Absolute, out Uri uri))
+                    {
+                        throw new SecurityTokenInvalidIssuerException("The token issuer is not valid.");
+                    }
+
+                    var tenant = _runningShellTable.Match(uri.Authority, uri.AbsolutePath);
```

Comment:
Note: this part must be reworked as it gives invalid results due to its fallback mechanism. I'm not yet sure what's the best way to fix that...

/cc @sebastienros

Reply:
We can change the API to optionally match explicitly

Reply:
That would be nice. Also, it needs to take the port into account.